contracts/bprotocol/scoring/BTokenScore.sol

```diff
@@ -33,48 +69,125 @@ contract BTokenScore is ScoringMachine {
     // =============
-    function updateDebtScore(address _user, address cToken, int256 amount) external onlyAvatar {
-        updateScore(user(_user), debtAsset(cToken), amount, now);
+    function updateDebtScore(address _avatar, address cToken, int256 amount) external onlyAvatar {
+        updateScore(user(_avatar), debtAsset(cToken), amount, now);
+    }
+
+    function updateCollScore(address _avatar, address cToken, int256 amount) external onlyAvatar {
+        updateScore(user(_avatar), collAsset(cToken), amount, now);
+    }
+
+    function mul_(uint score, uint multiplier, uint224 index) internal pure returns (uint) {
+        return mul_(mul_(score, multiplier), index);
     }
 
-    function updateCollScore(address _user, address cToken, int256 amount) external onlyAvatar {
-        updateScore(user(_user), collAsset(cToken), amount, now);
+    function updateIndex(address cToken) public {
+        require(expiry < now, "expired");
+        uint224 supplyIndex;
+        uint224 borrowIndex;
+        (supplyIndex,) = comptroller.compSupplyState(cToken);
+        (borrowIndex,) = comptroller.compBorrowState(cToken);
+        lastIndex[cToken].supplyIndex = supplyIndex;
+        lastIndex[cToken].borrowIndex = borrowIndex;
     }
 
-    function slashedScore(address _user, address cToken, int256 amount) external {
-        // TODO implement
-        // updateScore(user(_user), slashedAsset(cToken), amount, now);
+    function _getDeltaSupplyIndex(address cToken) internal returns (uint224 deltaSupplyIndex) {
+        uint224 currSupplyIndex;
+        (currSupplyIndex,) = comptroller.compSupplyState(cToken);
+        uint deltaSupplyIndexForCToken = sub_(uint(currSupplyIndex), uint(lastIndex[cToken].supplyIndex));
+
+        // NOTICE: supplyIndex takes cToken.totalSupply() which is in cToken quantity
+        // We need the index normalized to underlying token quantity
+        uint exchangeRate = ICToken(cToken).exchangeRateCurrent();
+        deltaSupplyIndex = safe224(div_(deltaSupplyIndexForCToken, exchangeRate), "index-exceeds-224-bits");
+    }
+
+    function _getDeltaBorrowIndex(address cToken) internal returns (uint224 deltaBorrowIndex) {
+        uint224 currBorrowIndex;
+        (currBorrowIndex,) = comptroller.compBorrowState(cToken);
+        deltaBorrowIndex = safe224(
+            sub_(uint(currBorrowIndex), uint(lastIndex[cToken].borrowIndex)),
+            "unable-to-cast-to-uint224"
+        );
+        // NOTICE: borrowIndex is already normalized to underlying token quantity
+    }
+
+    function slashedScore(address _avatar, address cToken) external {
+        IBComptroller bComptroller = IBComptroller(registry.bComptroller());
+        address _user = registry.ownerOf(_avatar);
+        IBToken bToken = IBToken(bComptroller.c2b(cToken));
+
+        uint time = sub(start, 30 days);
+        if(time < start) time = start;
+
+        // Slash debtScore
+        uint borrowBal_B = bToken.borrowBalanceCurrent(_user);
```

Comment:
need to make sure (test) it is not affected by cushion. also please avoid `_`. can just have `borrowBalB`.